### ISSUE TITLE
Make PyPi packages compliant with PEP 625

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#12181]
**Impact of the new feature**
WMCore in general

**Is your feature request related to a problem? Please describe.**
I just received the following email [1] reporting that some of our WMCore package names are not compliant with PEP 625, which will soon start being enforced by PyPi.

**Describe the solution you'd like**
From what I read, we need to replace `-` sign by `_` in the package names, for all packages uploaded to PyPi (global-workqueue, reqmgr2ms-pileup, etc etc).

**Describe alternatives you've considered**
No date was provided, but we better do it sooner than later.

**Additional context**
[1]
"""
This email is notifying you of an upcoming deprecation that we have determined may affect you as a result of your recent upload to 'reqmgr2ms-transferor'.

In the future, PyPI will require all newly uploaded source distribution filenames to comply with [PEP 625](https://peps.python.org/pep-0625/). Any source distributions already uploaded will remain in place as-is and do not need to be updated.

Specifically, your recent upload of 'reqmgr2ms-transferor-2.3.8rc5.tar.gz' is incompatible with PEP 625 because it does not contain the normalized project name 'reqmgr2ms_transferor'.

In most cases, this can be resolved by upgrading the version of your build tooling to a later version that supports PEP 625 and produces compliant filenames.
"""

**Comment by vkuznet** [comment]
From my understanding the steps which will be required to fix this issue are the following:
- change `requirements.txt` and replace `-` with `_` in WM specific package names, e.g. we need to change the following line:
```
pymongo~=4.2.0                # wmcore,wmagent-devtools,reqmgr2ms-unmerged,reqmgr2ms-output,reqmgr2ms-pileup
```
to
```
pymongo~=4.2.0                # wmcore,wmagent_devtools,reqmgr2ms_unmerged,reqmgr2ms_output,reqmgr2ms_pileup
```
- modify `.github/workflows/pypi_build_and_images.yaml` and adjust values in `target` list of strategy matrix, e.g.
```
        target: [wmagent, wmagent-devtools, wmcore, reqmon, reqmgr2, global-workqueue, acdcserver, reqmgr2ms-unmerged,
                 reqmgr2ms-output, reqmgr2ms-pileup, reqmgr2ms-rulecleaner, reqmgr2ms-transferor, reqmgr2ms-monitor]
```
to
```
        target: [wmagent, wmagent_devtools, wmcore, reqmon, reqmgr2, global_workqueue, acdcserver, reqmgr2ms_unmerged,
                 reqmgr2ms_output, reqmgr2ms_pileup, reqmgr2ms_rulecleaner, reqmgr2ms_transferor, reqmgr2ms_monitor]
```
- ensure that GitHub Action will properly build packages
- verify that new packages will be uploaded to PyPI
- change CMSKubernetes/docker/pypi areas to use underscores
- modify all `Dockerfile`s in CMSKubernetes/docker/pypi area to use new notations
- build new images with new names

@amaltaro, am I missing something?

**Comment by amaltaro** [comment]
Yes, I think you covered it well. In general, we have to cover this new convention in:
1) packaging and publishing to pypi
2) deploying it in Dockerfiles
3) deploying it in virtual environments
4) CMSKubernetes can likely remain as is - such that it does not impose any needs on extra work from the HTTP team

This can affect either manual or automatic mechanisms (e.g. GH pipelines).

In addition, I don't know if this file is used: https://github.com/dmwm/WMCore/blob/master/tools/build_pypi_packages.sh#L17, if it is, then we need to update it as well.
And perhaps this one too: https://github.com/dmwm/WMCore/blob/master/setup_dependencies.py

FYI @belforte @LinaresToine this might affect you as well

**Comment by belforte** [comment]
thanks Alan.  I think CRAB is not affected because we do not really use pypi but build our CRABServer on top of your docker image `registry.cern.ch/cmsweb/wmagent-base:pypi-20230705` . Maybe we should use a newer one, but that's a different story.  For TW and Publisher we simply start from Pytho3.8 image and pull code from GH.  "pip install crab_tw" was too much for our skills.

**Comment by vkuznet** [comment]
@amaltaro I found major issue with constrains on GH action and python package itself which will not allowed this transition. Full details of my investigation can be found in https://github.com/dmwm/WMCore/pull/12233#issuecomment-2602876873

**Comment by amaltaro** [comment]
Even though our new build system now relies on the setuptools library, we still have packages being published to PyPi with dashes `-`:
```
## same thing as in the original issue description, etc etc
Specifically, your recent upload of 'reqmgr2ms-monitor-2.3.10rc7.tar.gz' is incompatible with PEP 625 because the filename does not contain the normalized project name 'reqmgr2ms_monitor'. 
```

I had understood setuptools would have made that compliant for us, automatically. Clearly something is still missing and we will have to see which options we have - if possible, a solution that will be backwards compatible would be highly desirable.

**Comment by vkuznet** [comment]
Alan, I suggest that you re-read my https://github.com/dmwm/WMCore/pull/12233#issuecomment-2603046257 which shows inconsistencies within different Python PEPs proposals. I explicitly ping-point that Python internal function `.../site-packages/packaging/utils.py` leads to dashes instead of underscores based on a specific PEP proposal used in a specific python version.

**Comment by khurtado** [comment]
@vkuznet I agree with your comment here: https://github.com/dmwm/WMCore/pull/12233#issuecomment-2602876873

Given that the problem is basically that we currently depend on:

https://github.com/pypa/gh-action-pypi-publish

right [here](https://github.com/dmwm/WMCore/blob/master/.github/workflows/pypi_build_and_images.yaml#L97-L100)


How about we replace it with twine, directly? We do not use many of the features from the gh actions above anyway, and we want to get rid of the package renaming.

Twine does support trusted publishing since Jan 17, 2025:

https://github.com/pypa/twine/blob/0d0fc2dcf6445b3b93625978752408b3af958e77/docs/changelog.rst#twine-610-2025-01-17

So I picture something like this (this is just an example), should work:

```
      ...
        environment:
          name: production
        permissions:
          id-token: write
        run: |
          python -m pip install --upgrade twine
          twine upload --repository pypi dist/*
```

what do you think?

**Comment by vkuznet** [comment]
@khurtado , well `twine` is fine but still it is external tool which we can't control and therefore will rely on its compliance with PEP. For example, I can't say if it is complaint with PEP-625 or not as it itself can depend on many other packages.

That said, `twine` is upload tool and it will take whatever we create in `dist` area and upload it. As you can see our package archive (tar-balls or wheels) names in `dist` 
```
ls dist
reqmgr2ms_monitor-2.3.8-py3-none-any.whl
reqmgr2ms_monitor-2.3.8.tar.gz
```
Does contain combination of underscores and dashes and this **the problem** with respect to PEP-625. Therefore, replacing gh action with `twine` will not solve the underlying problem.

**Comment by vkuznet** [comment]
Hi, after our discussion during retreat meeting about using twine, I went ahead to check what the current GH Action is using, the one we have over here: https://github.com/dmwm/WMCore/blob/master/.github/workflows/pypi_build_and_images.yaml#L100

And, guess what I found :), they use twine, see https://github.com/pypa/gh-action-pypi-publish/blob/release/v1.12/twine-upload.sh. And, I found requirements runtime file https://github.com/pypa/gh-action-pypi-publish/blob/release/v1.12/requirements/runtime-constraints.in which specify twine version 6.1 which requires `packaging >= 24.2`. And, if you install package via pip you will find that

```
# Core metadata spec for `Name`
_validate_regex = re.compile(
    r"^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$", re.IGNORECASE
)
_canonicalize_regex = re.compile(r"[-_.]+")
_normalized_regex = re.compile(r"^([a-z0-9]|[a-z0-9]([a-z0-9-](?!--))*[a-z0-9])$")
# PEP 427: The build number must start with a digit.
_build_tag_regex = re.compile(r"(\d+)(.*)")


def canonicalize_name(name: str, *, validate: bool = False) -> NormalizedName:
    if validate and not _validate_regex.match(name):
        raise InvalidName(f"name is invalid: {name!r}")
    # This is taken from PEP 503.
    value = _canonicalize_regex.sub("-", name).lower()
    return cast(NormalizedName, value)
```
Which is what I originally reported in https://github.com/dmwm/WMCore/pull/12233#issuecomment-2603046257. It means that `twine` used in GitHub Action requires `packaging` which uses PEP 503 and not PEP 625. Maybe newer twine requires newer packaging which may be complain with PEP 625.

I suggest that either of you, maybe @khurtado , double check that. Otherwise, I think, simply dropping GH action and replacing with plain twine call does not solve the underlying issue.
